---
title: "Pseudobulk PCA Basal Analysis"
author: "Harvard Chan Bioinformatics Core"
date: "`r Sys.Date()`"
output:
   html_document:
      code_folding: hide
      df_print: paged
      highlights: pygments
      number_sections: false
      self_contained: true
      theme: default
      toc: true
      toc_float:
         collapsed: true
         smooth_scroll: true
editor_options: 
  chunk_output_type: console
params:
  project_file: information.R
  seurat_qs_fn: ../data_processed/DtST.qs
  meta_fn: ../meta/20240729_metadata_sc.xlsx
  agg_counts_fn: ../results/pseudobulk_counts_PCA_basal.csv
  agg_coldata_fn: ../results/pseudobulk_coldata_PCA_basal.csv
  
---

```{r load_params, echo = F}
source(params$project_file)
```

```{r load_libraries, cache = FALSE, message = FALSE, warning=FALSE, echo=FALSE,}

current_libs <- .libPaths()
.libPaths(c('/n/app/bcbio/R4.3.1_singlecell/', current_libs))
library(qs)
library(tidyverse)
library(Seurat)
# library(Matrix)
library(DESeq2)
library(DEGreport)
library(bcbioR)
library(ggprism)
library(knitr)
library(patchwork)
library(ggpubr)
library(pheatmap)
# library(BiocParallel)
# register(MulticoreParam(8))

options(stringsAsFactors = FALSE)

colors=cb_friendly_cols(1:15)
ggplot2::theme_set(theme_prism(base_size = 14))
opts_chunk[["set"]](
    cache = F,
    cache.lazy = FALSE,
    dev = c("png", "pdf"),
    error = TRUE,
    highlight = TRUE,
    message = FALSE,
    prompt = FALSE,
    tidy = FALSE,
    warning = FALSE,
    echo = T, 
    fig.height = 4)

# set seed for reproducibility
set.seed(1234567890L)

```

```{r sanitize_datatable}
sanitize_datatable = function(df, ...) {
 # remove dashes which cause wrapping
 DT::datatable(df, ..., rownames=gsub("-", "_", rownames(df)),
                   colnames=gsub("-", "_", colnames(df)),
               filter = 'top')
}
```

# Overview

-   Project: `r project`
-   PI: `r PI`
-   Analyst: `r analyst`
-   Experiment: `r experiment`
-   Aim: `r aim`

```{r load data, eval = !file.exists(params$agg_counts_fn)}

scrna <- qread('../data_processed/DtST.qs')
scrna_meta <- scrna@meta.data %>% as.data.frame()

ribo_genes <- grep(rownames(scrna), pattern="^RPS|^RPL")
scrna <- scrna[-ribo_genes,]

scrna$dtc <- factor(paste('d', scrna$donor, scrna$time, scrna$predicted.celltype.l1, sep = "_"))

meta <- readxl::read_xlsx(params$meta_fn) 

```

# Objective

Determine which covariates to include in differential expression modeling of non-stimulated, basal CSF samples and how best to use the covariates.  


```{r subset csf, eval = !file.exists(params$agg_counts_fn)}
csf <-subset(scrna, 
             cells=colnames(scrna)[
               which(
                 scrna$tissue == "csf" &
                   !(scrna$donor %in% c('242_003', '242_010', '242_014', '2577_010', 'LS1089')))
               ])

```

```{r aggregate counts csf, eval = !file.exists(params$agg_counts_fn)}

agg_counts_csf <- AggregateExpression(csf, assays = "RNA", group.by = c("dtc"))

agg_counts_csf <- agg_counts_csf$RNA
colnames(agg_counts_csf) <- sapply(colnames(agg_counts_csf), function(x) gsub('-', '_', x)) %>% as.character()

csf_meta <- csf@meta.data %>% as.data.frame()

coldata_csf <- csf@meta.data %>% as.data.frame() %>% 
  dplyr::select(donor, time, predicted.celltype.l1, dtc) %>%
  distinct() %>%
  mutate(dt = paste(donor, time, sep ='_')) %>%
  left_join(meta %>% mutate(donor = gsub('-', '_', subject),
                            sex1m = as.factor(sex1m), 
                            apoe4_carrier = as.factor(apoe4_carrier),
                            disease = ifelse(km_ptau181_div_ab42, 'AD', 'Healthy')) %>% 
              select(donor, age, apoe4_carrier, sex1m, disease) %>% distinct())

coldata_csf$dtc <- sapply(coldata_csf$dtc, function(x) gsub('-', '_', x))
rownames(coldata_csf) <- coldata_csf$dtc
coldata_csf <- coldata_csf[colnames(agg_counts_csf), ]
stopifnot(all(rownames(coldata_csf) == colnames(agg_counts_csf)))

dds_csf <- DESeqDataSetFromMatrix(agg_counts_csf,
                                  colData = coldata_csf,
                                  design = ~1)
dds_csf <- DESeq(dds_csf)
  
vst_csf <- vst(dds_csf)

vst_csf_df <- assay(vst_csf) %>% as.data.frame()
rownames(vst_csf_df) <- rownames(agg_counts_csf)

write_csv(vst_csf_df %>% rownames_to_column('gene_name'), params$agg_counts_fn)
write_csv(coldata_csf, params$agg_coldata_fn)

```


# PCA

The largest cause of variability between the psuedobulked samples is cell type. After that, the largest causes of variability are donor and disease status. 

```{r read data}
vst_csf_df <- read_csv(params$agg_counts_fn) %>% column_to_rownames('gene_name')
coldata_csf <- read_csv(params$agg_coldata_fn) 
rownames(coldata_csf) <- coldata_csf$dtc

```

## All

```{r pca all csf}

degPCA(as.matrix(vst_csf_df), coldata_csf %>% as.matrix(), condition = 'predicted.celltype.l1')

```

## By Cell Type {.tabset}

```{r cell type csf, results = 'asis', fig.height = 6, fig.width = 10}

for (cell_type in unique(coldata_csf$predicted.celltype.l1)){
  
  cat("### ", cell_type, "\n")

  coldata_csf_cell_type <- coldata_csf %>% filter(predicted.celltype.l1 == cell_type)
  vst_csf_df_cell_type <- vst_csf_df[, rownames(coldata_csf_cell_type)]
  
  p1 <- degPCA(as.matrix(vst_csf_df_cell_type), coldata_csf_cell_type %>% as.matrix(), condition = 'donor', shape = 'time')
  p2 <- degPCA(as.matrix(vst_csf_df_cell_type), coldata_csf_cell_type %>% as.matrix(), condition = 'disease', shape = 'time')
  
  p <- wrap_plots(p1, p2, ncol = 2)
  print(p)
  
  vst_cell_type_cor <- cor(as.matrix(vst_csf_df_cell_type))

  colma=coldata_csf_cell_type %>% as.data.frame() %>% mutate(time = factor(time, levels = c(0, 3, 12)))
  rownames(colma) <- colma$dtc
  colma <- colma[rownames(vst_cell_type_cor), ]
  colma <- colma %>% dplyr::select(disease, donor, time)
  anno_colors=lapply(colnames(colma), function(c){
    l.col=cb_friendly_pal('main')(length(unique(colma[[c]])))
    names(l.col)=unique(colma[[c]])
    l.col
  })
  names(anno_colors)=colnames(colma)
  
  p <- pheatmap(vst_cell_type_cor,
           annotation = colma,
           annotation_colors = anno_colors,
           show_rownames = F,
           show_colnames = F,
           color = cb_friendly_pal('heatmap')(15)
           )
  p
  
  
  
  cat('\n\n')
}

```

# Evaluate Other Covariates

For this plot, if the point is outlined, the correlation between that variable and PC has a FDR < 0.05. A point being colored means that its correlation is significant even when considered among other covariates with outlined points for that PC. 

In this case, we can see that donor, dt (donor x time), and cell type are individually significantly correlated with PC3, but only cell type remains significantly correlated with PC1 when all 3 covariates are modeled together.  In the heatmap, we see that some covariates are strongly correlated (or anti-correlated) with each other, but in ways that were expected (e.g. donor correlated with donor x time). 


```{r covariates csf}

coldata_csf_complete <- coldata_csf %>% drop_na()
rownames(coldata_csf_complete) <- coldata_csf_complete$dtc
vst_csf_df_complete <- vst_csf_df[, rownames(coldata_csf_complete)]
p <- degCovariates(as.matrix(vst_csf_df_complete), coldata_csf_complete %>% as.matrix(), plot = F)

p$plot
```


```{r covariates cor csf}
degCorCov(coldata_csf_complete)
```

# Conclusions

These plots help us determine what factors should be considered in an eventual differential expression model. Since we are interested in the effect of time on gene expression, we obviously must include it as a factor. 

Additionally, we see that cell type and donor also heavily influence the variability of the pseudobulked samples. More specifically, since the amount of variance between samples heavily depends on cell type (e.g. CD4 T cell pseudobulk samples are considerably more variable than NK pseudobulk samples), best practice is to subset the samples by cell type and fit one model per cell type. 

Since the amount of variance between samples does not heavily depend on donor (e.g. pseudobulked samples from donor 242_003 are not considerably more variable than pseudobulked samples from donor 242_010), or disease (e.g. samples from patients with AD are not significantly more variable than samples from healthy patients) it is acceptable to merely include donor or disease as a factor in the model as opposed to having separate models for each value of donor and/or disease. Unfortunately, it is not possible to model both donor and disease status together--it is impossible to distinguish the effect of donor from the effect of disease because one group of donors has AD and an entirely separate group of donors is healthy. 

Finally, we examine additional metadata factors such as age, apoe3 status, and sex, and find that they account for a vanishingly small proportion of the differences between samples, meaning that they can be reasonably excluded from the model. 

